Add CONTRIBUTING.md (#1758)
functional-streams-for-scala/fs2: CONTRIBUTING.md

Remember to follow the [code of conduct][coc] in online and offline discourse.

## Building the project locally

### Prerequisites

You'll need JDK 11 and [sbt][sbt].

We use several sbt plugins to build and check the project, including [MiMa (Migration Manager)][mima], [scalafmt][scalafmt] and [sbt-microsites][sbt-microsites].

### Build process

To compile the code for the whole repository, you can start an interactive sbt shell:

```bash
$ sbt
[info] Loading global plugins from /Users/contributor/.sbt/1.0/plugins
[info] Loading settings for project fs2-build from plugins.sbt,metals.sbt ...
[info] Loading project definition from /Users/contributor/dev/fs2/project
[info] Loading settings for project root from version.sbt,build.sbt ...
[info] Set current project to root (in build file:/Users/contributor/dev/fs2/)
sbt:root>
```

Inside the shell, you can compile the sources for the currently selected Scala version using the `compile` command.
To compile the code for all Scala versions enabled in the build, use `+compile`. To include tests, `test:compile` or `+test:compile`, accordingly.

### Testing

> To test the code, you can run the `testJVM` or `testJS` command, depending on the platform you want to run the tests on.

Reviewer: are there commands to run _all_ tests, or do you _have_ to run `testJVM` _and_ `testJS`?

Author: You can just do `test`. I'll mention that...

Before submitting a change for review, it's worth running some extra checks that will be triggered in Continuous Integration:

```sbt
sbt:root> fmtCheck; testJVM; testJS; doc; mimaReportBinaryIssues; docs/mdoc; microsite/makeMicrosite
```

That will check the formatting, run all tests on the JVM and JS platforms, report any binary compatibility issues (as detected by [MiMa][mima]) and build the site.

If you run into any problems with tests, binary compatibility or other issues, feel free to ask questions on [the Gitter channel][fs2-dev].

### Website

To build the microsite, use the `microsite/makeMicrosite` task in sbt. Then, go to `site/target/site`, and run a HTTP server on the port of your choice:

```bash
sbt:root> microsite/makeMicrosite
info: Compiled in 24s (0 errors, 0 warnings)

> [info] Configuration file: /Users/contributor/dev/fs2/site/target/scala-2.13/resource_managed/main/jekyll/_config.yml

Reviewer: are any other developer requirements necessary (ruby / jekyll / etc)?

Curious as I have everything on my laptop but not sure what else is needed for microsite

Author: Yeah, pretty sure we'll also need jekyll to build the microsite. I'll mention that.

Author: I'll try building it in docker on a clean image and see what else is missing :)

Author: okay, basically we were missing jekyll and node (for running scalajs).

sbt:root>
```

Inside the shell, you can compile the sources for the currently selected Scala version using the `compile` command.
To compile the code for all Scala versions enabled in the build, use `+compile`. To include tests, `test:compile` or `+test:compile`, accordingly.

### Testing

To test the code, you can run the `test` command in sbt.
If you want the tests on a single plaform, you can use `testJVM` or `testJS` instead (these are defined as aliases in the `build.sbt` file).

It is possible to run a single test suite from a project on a single platform by [executing a more specific task](https://www.scala-sbt.org/1.x/docs/Testing.html#testOnly), like `coreJVM/testOnly fs2.PullSpec`.

You can list all available projects by executing the `projects` task:

```sbt
sbt:root> projects
[info] In file:/Users/kubukoz/IdeaProjects/fs2/
[info] 	   benchmark
[info] 	   coreJS
[info] 	   coreJVM
[info] 	   docs
[info] 	   experimental
[info] 	   io
[info] 	   microsite
[info] 	   reactiveStreams
[info] 	 * root
```

Before submitting a change for review, it's worth running some extra checks that will be triggered in Continuous Integration:

```sbt

> sbt:root> fmtCheck; testJVM; testJS; doc; mimaReportBinaryIssues; docs/mdoc; microsite/makeMicrosite

Author: ```suggestion
sbt:root> fmtCheck; test; doc; mimaReportBinaryIssues; docs/mdoc; microsite/makeMicrosite
```

```

> That will check the formatting, run all tests on the JVM and JS platforms, report any binary compatibility issues (as detected by [MiMa][mima]) and build the site.

Author: ```suggestion
That will check the formatting, run all tests on the supported platforms, report any binary compatibility issues (as detected by [MiMa][mima]) and build the site.
```

If you run into any problems with tests, binary compatibility or other issues, feel free to ask questions on [the Gitter channel][fs2-dev].

### Website

To build the microsite, use the `microsite/makeMicrosite` task in sbt. Then, go to `site/target/site`, and run a HTTP server on the port of your choice:

```bash
sbt:root> microsite/makeMicrosite
info: Compiled in 24s (0 errors, 0 warnings)
[info] Configuration file: /Users/contributor/dev/fs2/site/target/scala-2.13/resource_managed/main/jekyll/_config.yml
[info]             Source: /Users/contributor/dev/fs2/site/target/scala-2.13/resource_managed/main/jekyll
[info]        Destination: /Users/contributor/dev/fs2/site/target/jekyll
[info]  Incremental build: disabled. Enable with --incremental
[info]       Generating...
[info]                     done in 0.592 seconds.
[info]  Auto-regeneration: disabled. Use --watch to enable.
[success] Total time: 29 s, completed Jan 24, 2020, 2:36:31 PM
sbt:root> exit
[info] shutting down sbt server

$ cd site/target/site
$ python -m SimpleHTTPServer 4000 . # Any other server would do

> Serving HTTP on 0.0.0.0 port 8080 ...

Author: Port doesn't match command.

> Serving HTTP on 0.0.0.0 port 8080 ...

Author: ```suggestion
Serving HTTP on 0.0.0.0 port 4000 ...
```
